cpu: BPU support for decoupled front-end

dhschall opened the pull request:
To support the decoupled front-end the BPU must have the ability to add a placeholder for branches that are not in
the BTB (surprise branches). In this case the front-end cannot detect the branch and make a prediction. However,
to rollback in case of a misprediction the BPU needs a history element. This function allows generation of such elements to add it to the the speculative history buffer. It does NOT make a prediction or alter the BPU state.

Harshil2107 commented:
Hello @andysan ,
Would it be alright if I assign this PR to you? Please be aware that being assigned to a PR implies that you are responsible for ensuring that the PR is reviewed and merged.

rjc-arch commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -187,6 +187,21 @@ class BPredUnit : public SimObject
+    /**
+     * Special function for the decoupled front-end. In it there can be
+     * branches which are not detected by the BPU in the first place as it
+     * requires a BTB hit. This function will generage a placeholder for
```
Typo: 'generage' -> 'generate'.

rjc-arch commented on src/cpu/pred/bpred_unit.cc:
```diff
@@ -93,6 +93,12 @@ BPredUnit::drainSanityCheck() const
+BPredUnit::branchPlaceholder(ThreadID tid, Addr pc,
+                             bool uncond, void * &bp_history)
+{
+    panic("Not implemented for this BP!\n");
```
Optional: you could add a bit more context to the message here to avoid the user  having to look in the source code for the reason. For example "branchPlaceholder() not implemented for this BP." .

rjc-arch commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -187,6 +187,21 @@ class BPredUnit : public SimObject
+     * branches which are not detected by the BPU in the first place as it
+     * requires a BTB hit. This function will generage a placeholder for
+     * such a branch once it is pre-decoded in the fetch stage. It will
+     * only create the branch history object but not update BPU internal.
```
Typo(?): "BPU internal" -> "the BPU internal state"?

rjc-arch commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -187,6 +187,21 @@ class BPredUnit : public SimObject
+     * requires a BTB hit. This function will generage a placeholder for
+     * such a branch once it is pre-decoded in the fetch stage. It will
+     * only create the branch history object but not update BPU internal.
+     * If the branch turns to be wrong then decode or commit will
```
Typo 'user' -> 'use'.
"If the branch turns out to be wrong then decode or commit will
be able to use the normal squash functionality..."

rjc-arch commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -187,6 +187,21 @@ class BPredUnit : public SimObject
+     * If the branch turns to be wrong then decode or commit will
+     * be able to user the normal squash functionality to correct the branch.
+     * Note that not all branch predictors implement this functionality.
+     * @param PC The branch's PC.
```
Nit: 'PC' -> 'pc'.

rjc-arch commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -187,6 +187,21 @@ class BPredUnit : public SimObject
+     * be able to user the normal squash functionality to correct the branch.
+     * Note that not all branch predictors implement this functionality.
+     * @param PC The branch's PC.
+     * @param uncond Wheather or not this branch is an unconditional branch.
```
Typo: 'wheather' -> 'whether'.

rjc-arch commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -348,7 +363,8 @@ class BPredUnit : public SimObject
 
     /**
-     * Squashes a particular branch instance
+     * Squashes a particular branch instance. Reverts
```
Optional suggestion: since this patch changes where the history items are deleted, you could add a comment here to explain which method is responsible for deleting the history items. 
Although it's not part of your code changes, I've found the memory management of the BP history, etc, quite complex. Some documentation might not go amiss.

dhschall replied:
@rjc-arch thanks for the review! I've added more details on the PredictorHistory object which hopefully helps to better understand its live-cycle

rjc-arch replied:
Thanks for the extra comments - these are very helpful!

ivanaamit commented:
Just checking in -- what is the status of this PR? Thanks.

rjc-arch commented on src/cpu/pred/tage_base.cc:
```diff
@@ -319,21 +300,41 @@ TAGEBase::baseUpdate(Addr pc, bool taken, BranchInfo* bi)
-             tab[histBufferSize - maxHist + i] = tab[i];
-         pt =  histBufferSize - maxHist;
-         h = &tab[pt];
+            tHist.globalHistory[histBufferSize - maxHist + i]
```
Nit: For clarity I would put braces around the for loop body since it is split over two lines.

rjc-arch commented on src/cpu/pred/tage_base.cc:
```diff
@@ -319,21 +300,41 @@ TAGEBase::baseUpdate(Addr pc, bool taken, BranchInfo* bi)
         DPRINTF(Tage, "Rolling over the histories\n");
          // Copy beginning of globalHistoryBuffer to end, such that
          // the last maxHist outcomes are still reachable
          // through pt[0 .. maxHist - 1].
```
`pt` has been removed now.

rjc-arch commented on src/cpu/pred/tage_base.hh:
```diff
@@ -279,23 +300,53 @@ class TAGEBase : public SimObject
+     * @param tid The thread ID to select the histories to update.
+     * @param brtype The branch type
+     * @param taken Actual branch outcome.
+     * @param pc The unshifted branch PC.
```
`pc` has been re-named to `branch_pc`.

rjc-arch reviewed (changes requested):
Hi @dhschall - this all looks good to me. I'm not very familiar with the details of the branch predictor implementations, but I can't see any obvious problems.

I have commented on some small, mostly cosmetic changes you might consider. There are also a number of places with incorrect/inconsistent indentation/formatting that I haven't commented on, but which you might also like to review (mostly in the `tage_base.(hh|cc)` files).

Unless we can find someone with a deeper knowledge of the branch predictors to do a review, I think this looks good.

rjc-arch commented:
We have also run some performance regressions of this PR using SPEC 2017 SimPoints, and have confirmed that it does not cause any unexpected changes in simulated performance.

dhschall committed:
cpu: BPU support for surprise branches

To support the decoupled front-end the BPU must have the
ability to add a placeholder for branches that are not in
the BTB (surprise branches). In this case the front-end
cannot detect the branch and make a prediction. However,
to roleback in case of a misprediction the BPU needs a
history element.
This function allows the to add such element after pre-
decode. It does NOT make a prediction or alter the BPU
state.
TAGE's global history management is update to support
roleback of surprise branches.

Change-Id: I4446f5f29908d75315ecde1eca26986b21324896
Signed-off-by: David Schall <david.schall@ed.ac.uk>

Change-Id: Ic1ab3f0320ec17b923b0ac23eddf187a58649259

erin-le commented:
@dhschall Hello, thank you for your contribution! The gem5 dev team would like to help you get this PR in, but we have a few questions about this and your other PRs. 

First, could you clarify how this PR, https://github.com/gem5/gem5/pull/359, https://github.com/gem5/gem5/pull/1854, https://github.com/gem5/gem5/pull/1855 fit together? 

Also, how should these PRs be tested, i.e. should they all be cherry-picked onto one branch and tested together, tested completely separately, or something in between?  Furthermore, how have you been testing these PRs?

Once the gem5 dev team has a better understanding of these PRs and how they should be tested, we can help to develop a plan for testing these changes and getting them into gem5. Thank you!

dhschall commented:
Thank you @erin-le for the quick feedback!
This PR is the basis for all the other PR's mentioned. It is mostly refactoring of the branch predictor to make the other changes possible/easier. Furthermore, it enables adding _surprise-branches_, branches that were not detected by the front-end due to BTB misses. This will become relevant for the decoupled front-end later.
As for now, this PR should not change the behavior. @rjc-arch already did some regression tests with SPEC workloads a while ago, and the behavior didn't change then.

#1854 Adds support for speculatively updating TAGE-SC-L (see #1341 for the discussion). It is independent of the other changes.
#1855 Adds support for taken-only history which is needed for the decoupled front-end.
#359 Is the implementation of the decoupled front-end. However, it is not mergeable as the cache interface changes. Also, I did some significant refactoring, but I need a bit more time to finalize and test it. #359 depends on this PR and #1855.

For testing, I am using a few more complex server workloads in full system simulations (see [graph](https://github.com/user-attachments/assets/919ed4c8-dd55-4976-bd5b-c0500bf24d44) in #1854). One workload is a NodeJS web server running inside a docker container that is invoked by a client running on a second node. Two other workloads (libs and compression) are taken from Google's fleet-bench benchmark suite.
I'll run again the tests for this PR and add the graph. If you have some testing infrastructure or some workloads, you want me to run, let me know. I've also asked @rjc-arch to run SPEC simpoints again, but he is quite busy atm. so he might not be able to help me with further testing.

powerjg commented:
Thanks for all of the details, David! I think it would be good for us here at UCD to run some SPEC simpoints and publish the data. We'll get on that.

It sounds like this PR is ready to go, is that right?

dhschall commented:
Thanks, @powerjg! Do you mean publishing the simpoints as executables, runnable on gem5, or that you publish the performance numbers? The former would be awesome to establish a common baseline for testing and validation. 
I don't know how the testing works but if you could integrate it with the CI to run these tests from time to time to observe how the simulated performance changes over time that would also be great but I know it takes time...

FE PR, from my side, yes. I've fixed some minor indentation issues and pre-commit checks.

powerjg commented:
I mean to publish the results from SPEC simpoints. Annoyingly, the SPEC license doesn't allow us to distribute the simpoints.

We could potentially integrate this into the testing infrastructure, since we control it. The main reason we haven't is the amount of time/resources that the simulation takes. Maybe we can integrate it into the weekly tests, though. We'll give it some thought!

rjc-arch commented on src/cpu/pred/2bit_local.cc:
```diff
@@ -79,8 +79,8 @@ LocalBP::LocalBP(const LocalBPParams &params)
-LocalBP::updateHistories(ThreadID tid, Addr pc, bool uncond,
-                         bool taken, Addr target, void * &bp_history)
+LocalBP::updateHistories(ThreadID tid, Addr pc, bool uncond, bool taken,
+                   Addr target, const StaticInstPtr &inst, void * &bp_history)
```
Nit: Code formatting: the alignment is slightly off on some of these function definitions. (The same applies to some other files too.)

rjc-arch commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -187,6 +190,22 @@ class BPredUnit : public SimObject
+     * If the branch turns to be wrong then decode or commit will
+     * be able to use the normal squash functionality to correct the branch.
+     * Note that not all branch predictors implement this functionality.
+     * @param pc The branch's PC.
```
Nit: The `tid` Thread ID parameter should also be listed here.

rjc-arch commented on src/cpu/pred/tage_base.hh:
```diff
@@ -110,6 +110,13 @@ class TAGEBase : public SimObject
+        {
+            comp ^= h[origLength] << outpoint;
+            auto tmp = (comp & 1) ^ h[0];
+            comp = ( tmp << (compLength-1)) | (comp >> 1);
```
Nit: there's an extra space after the opening brace.

rjc-arch commented on src/cpu/pred/tage_base.hh:
```diff
@@ -279,23 +304,53 @@ class TAGEBase : public SimObject
+    /**
+     * Records the current state of the histories to be able to restore it
+     * in case of a mispredicted speculative update.
+     * @param bi Pointer to the branch associated with the state
```
Nit: missing the `tid` parameter.

rjc-arch commented on src/cpu/pred/tage_base.hh:
```diff
@@ -279,23 +304,53 @@ class TAGEBase : public SimObject
+    /**
+     * Restore the state of the histories in case of detecting
+     * a mispredicted speculative update.
+     * @param bi Pointer to the branch associated with the state
```
Nit: missing the `tid` parameter.

rjc-arch commented on src/cpu/pred/tage_base.hh:
```diff
@@ -279,23 +304,53 @@ class TAGEBase : public SimObject
+     * @param taken Actual branch outcome.
+     * @param branch_pc The unshifted branch PC.
+     * @param target The branch target
+     */
```
Nit: missing the `bi` parameter.

rjc-arch commented on src/cpu/pred/tage_base.cc:
```diff
@@ -319,21 +300,42 @@ TAGEBase::baseUpdate(Addr pc, bool taken, BranchInfo* bi)
+        // Shift the next bit of the bit vector into the history
+        tHist.ptGhist--;
+        tHist.gHist--;
+        tHist.gHist[0] = (bv & 1) ? 1 : 0;
```
The syntax `*(tHist.gHist) = (bv & 1) ? 1 : 0;` might be slightly clearer here, to avoid making `tHist.gHist` look like an array. That's just a subjective opinion though, so feel free to ignore this comment if you don't agree.

rjc-arch reviewed (approved):
Hi @dhschall, 

Thank you for for the updates! This PR looks good to me. 

I have made a few comments on some very small issues like code formatting which you could fix, but I'd be happy for this PR to be merged.

We have also re-run the performance regressions mentioned above, and have re-confirmed that it does not cause any unexpected changes in simulated performance.

dhschall commented:
Thanks @rjc-arch! I've addressed all your formating suggestions. @andysan can you maybe have a second if its all right to get it merged?
